### PR TITLE
[Spark] Relax protocol reader feature requirements to allow backward compatibility

### PR DESCRIPTION
**Pull request by andreaschat-db**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

This is a followup of https://github.com/delta-io/delta/commit/920f185a04382ff466e47e75240dfda48efe40d3. We relax the protocol reader feature requirements to allow backward compatibility with older writers. This is because before https://github.com/delta-io/delta/commit/920f185a04382ff466e47e75240dfda48efe40d3 was possible to serialize invalid (2, 7) protocols without the reader features set.

## How was this patch tested?

Added new UTs.

## Does this PR introduce _any_ user-facing changes?

No.

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala [hunk 1]
```diff
@@ -139,10 +139,15 @@ case class Protocol private (
   extends Action
   with AbstractProtocol
   with TableFeatureSupport {
-  // Correctness check
-  // Reader and writer versions must match the status of reader and writer features
+  // Correctness check.
+  // Reader and writer versions must match the status of reader and writer features.
+  // For the reader features requirement we exclude the column mapping reader version (2)
```
Comment (tomvanbussel) [inline spark/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala:144]: This comment describes the case when the reader version is 3, but the reader features are not defined. The code below on the other hand handles the case in which the reader version is 2.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala [hunk 2]
```diff
@@ -2397,6 +2397,60 @@ trait DeltaProtocolVersionSuiteBase extends QueryTest
     }
   }
 
+  test("Can construct a protocol with column mapping only in writer features") {
+    // Column mapping should appear in both reader and writer features. However,
+    // we allow the construction of invalid protocols such as the one below to allow backward
```
Comment (tomvanbussel) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala:2402]: Why would the one below be invalid? Is it because `readerFeatures` should have been `None`?

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala [hunk 3]
```diff
@@ -2397,6 +2397,60 @@ trait DeltaProtocolVersionSuiteBase extends QueryTest
...
+      readerFeatures = Some(Set.empty),
+      writerFeatures = Some(Set(ColumnMappingTableFeature.name)))
+
+    Protocol(
+      minReaderVersion = 2,
+      minWriterVersion = 7,
+      readerFeatures = Some(Set.empty),
+      writerFeatures = Some(Set(ColumnMappingTableFeature.name, TestWriterFeature.name)))
+
+    // Valid protocol are also allowed.
```
Comment (tomvanbussel) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala:2418]: Surprised that this is valid. Shouldn't the reader features be missing in this case, because the reader protocol version does not support reader features?

**Review by tomvanbussel — APPROVED** [review]
Left a few questions, code changes LGTM otherwise.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala (hunk 3 above)
Comment (tomvanbussel) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala:2418]: Yes, but wouldn't reader version 2 already indicate that this table uses column mapping? And on the other hand reader features are not guaranteed to be understood when the reader version is 2.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala (hunk 2 above)
Comment (tomvanbussel) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala:2402]: But the reader version below is 2. In that case the readers features should not be set, right?

Comment (andreaschat-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala:2402]: Both (2, 5) and (2, 7, columnMapping, ...) are valid protocols. The first is in legacy form and the later in the table features form.

Once we are in the table features territory all features need to be explicitly set. In a sense, yes (2, 7) implies there is column mapping only because column mapping is the only reader feature with 2. However, the semantics of the table feature protocols is to explicitly state all supported features.

I will add an extra tests to also include the none case.
